service/history/mutableStateDecisionTaskManager.go

```diff
@@ -0,0 +1,726 @@
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/errors"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/persistence"
+)
+
+type (
+	mutableStateDecisionTaskManager interface {
+		ReplicateDecisionTaskScheduledEvent(
+			version int64,
+			scheduleID int64,
+			taskList string,
+			startToCloseTimeoutSeconds int32,
+			attempt int64,
+			scheduleTimestamp int64,
+			originalScheduledTimestamp int64,
+		) (*decisionInfo, error)
+		ReplicateTransientDecisionTaskScheduled() (*decisionInfo, error)
+		ReplicateDecisionTaskFailedEvent() error
+		ReplicateDecisionTaskTimedOutEvent(timeoutType workflow.TimeoutType) error
+		ReplicateDecisionTaskCompletedEvent(event *workflow.HistoryEvent) error
+		ReplicateDecisionTaskStartedEvent(
+			decision *decisionInfo,
+			version int64,
+			scheduleID int64,
+			startedID int64,
+			requestID string,
+			timestamp int64,
+		) (*decisionInfo, error)
+
+		AddDecisionTaskFailedEvent(
+			scheduleEventID int64,
+			startedEventID int64,
+			cause workflow.DecisionTaskFailedCause,
+			details []byte,
+			identity string,
+			reason string,
+			baseRunID string,
+			newRunID string,
+			forkEventVersion int64,
+		) (*workflow.HistoryEvent, error)
+		AddDecisionTaskScheduleToStartTimeoutEvent(scheduleEventID int64) (*workflow.HistoryEvent, error)
+		AddDecisionTaskTimedOutEvent(scheduleEventID int64, startedEventID int64) (*workflow.HistoryEvent, error)
+		AddDecisionTaskCompletedEvent(
+			scheduleEventID int64,
+			startedEventID int64,
+			request *workflow.RespondDecisionTaskCompletedRequest,
+			maxResetPoints int,
+		) (*workflow.HistoryEvent, error)
+		AddDecisionTaskStartedEvent(
+			scheduleEventID int64,
+			requestID string,
+			request *workflow.PollForDecisionTaskRequest,
+		) (*workflow.HistoryEvent, *decisionInfo, error)
+		AddDecisionTaskScheduledEventAsHeartbeat(
+			bypassTaskGeneration bool,
+			originalScheduledTimestamp int64,
+		) (*decisionInfo, error)
+		AddDecisionTaskScheduledEvent(bypassTaskGeneration bool) (*decisionInfo, error)
+		AddFirstDecisionTaskScheduled(startEvent *workflow.HistoryEvent) error
+		FailDecision(incrementAttempt bool)
+		DeleteDecision()
+		UpdateDecision(decision *decisionInfo)
+
+		HasPendingDecision() bool
+		GetPendingDecision() (*decisionInfo, bool)
+		HasInFlightDecision() bool
+		GetInFlightDecision() (*decisionInfo, bool)
+		HasProcessedOrPendingDecision() bool
+		GetDecisionInfo(scheduleEventID int64) (*decisionInfo, bool)
+
+		CreateTransientDecisionEvents(decision *decisionInfo, identity string) (*workflow.HistoryEvent, *workflow.HistoryEvent)
+	}
+
+	mutableStateDecisionTaskManagerImpl struct {
+		msb *mutableStateBuilder
+	}
+)
+
+func newMutableStateDecisionTaskManager(msb *mutableStateBuilder) mutableStateDecisionTaskManager {
+	return &mutableStateDecisionTaskManagerImpl{
+		msb: msb,
+	}
+}
+
+func (m *mutableStateDecisionTaskManagerImpl) ReplicateDecisionTaskScheduledEvent(
```

Comment:
could you plz arrange the functions in this way?

1. decision schedule functions, i.e. add then replicate
2. decision start functions
3. decision completed functions
4. decision timeout / fail, etc

Reply:
I reordered in this way but I kept the replicate methods separate.